File: dsa-problems/GraphsBasics/solutions/cpp/GRB-014-shortest-path-dag.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <algorithm>

using namespace std;

class Solution {
    vector<bool> visited;
    stack<int> st;

    void dfs(int u, const vector<vector<pair<int, int>>>& adj) {
        visited[u] = true;
        for (auto& edge : adj[u]) {
            int v = edge.first;
            if (!visited[v]) dfs(v, adj);
        }
        st.push(u);
    }

public:
    vector<long long> shortestPathDAG(int n, const vector<vector<pair<int, int>>>& adj, int s) {
        visited.assign(n, false);
        
        for (int i = 0; i < n; i++) {
            if (!visited[i]) dfs(i, adj);
        }

        vector<long long> dist(n, 2e18); // Large value for infinity
        dist[s] = 0;

        while (!st.empty()) {
            int u = st.top();
            st.pop();

            if (dist[u] != 2e18) {
                for (auto& edge : adj[u]) {
                    int v = edge.first;
                    int w = edge.second;
                    if (dist[u] + w < dist[v]) {
                        dist[v] = dist[u] + w;
                    }
                }
            }
        }

        for (int i = 0; i < n; i++) {
            if (dist[i] == 2e18) dist[i] = -1;
        }
        return dist;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m, s;
    if (!(cin >> n >> m >> s)) return 0;
    vector<vector<pair<int, int>>> adj(n);
    for (int i = 0; i < m; i++) {
        int u, v, w;
        cin >> u >> v >> w;
        adj[u].push_back({v, w});
    }

    Solution solution;
    vector<long long> dist = solution.shortestPathDAG(n, adj, s);
    for (int i = 0; i < n; i++) {
        if (i) cout << ' ';
        cout << dist[i];
    }
    return 0;
}

File: dsa-problems/GraphsBasics/solutions/java/GRB-014-shortest-path-dag.java
import java.util.*;

class Solution {
    private Stack<Integer> stack;
    private boolean[] visited;

    public long[] shortestPathDAG(int n, List<List<int[]>> adj, int s) {
        stack = new Stack<>();
        visited = new boolean[n];

        // 1. Topological Sort
        for (int i = 0; i < n; i++) {
            if (!visited[i]) dfs(i, adj);
        }

        // 2. Initialize Distances
        long[] dist = new long[n];
        Arrays.fill(dist, Long.MAX_VALUE);
        dist[s] = 0;

        // 3. Relax in Topological Order
        while (!stack.isEmpty()) {
            int u = stack.pop();
            
            if (dist[u] != Long.MAX_VALUE) {
                for (int[] edge : adj.get(u)) {
                    int v = edge[0];
                    int w = edge[1];
                    if (dist[u] + w < dist[v]) {
                        dist[v] = dist[u] + w;
                    }
                }
            }
        }

        // Convert unreachable to -1
        for (int i = 0; i < n; i++) {
            if (dist[i] == Long.MAX_VALUE) dist[i] = -1;
        }
        
        return dist;
    }

    private void dfs(int u, List<List<int[]>> adj) {
        visited[u] = true;
        for (int[] edge : adj.get(u)) {
            int v = edge[0];
            if (!visited[v]) dfs(v, adj);
        }
        stack.push(u);
    }
}

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if (!sc.hasNextInt()) return;
        int n = sc.nextInt();
        int m = sc.nextInt();
        int s = sc.nextInt();
        List<List<int[]>> adj = new ArrayList<>();
        for (int i = 0; i < n; i++) adj.add(new ArrayList<>());
        for (int i = 0; i < m; i++) {
            int u = sc.nextInt();
            int v = sc.nextInt();
            int w = sc.nextInt();
            adj.get(u).add(new int[]{v, w});
        }

        Solution solution = new Solution();
        long[] dist = solution.shortestPathDAG(n, adj, s);
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < n; i++) {
            if (i > 0) sb.append(' ');
            sb.append(dist[i]);
        }
        System.out.print(sb.toString());
        sc.close();
    }
}
